File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.cpp
class Solution {
public:
    string reverseWords(string s) {
        size_t i = 0, j = 0;   // (i -> start, j -> end) of each word
        const size_t& size = s.size();

        for (size_t k = 1; k < size; ++k) {
            if (s[k] == ' ') {
                // whitespace reached, start swapping
                swap_char(s, i, j);
                // update i, j to start index of next word
                i = j = k + 1;
            } else {
                // other ascii char reached, update j
                j = k;
            }
        }
        // no trailing spaces, manually swap last word
        swap_char(s, i, j);

        return s;
    }
private:
    inline void swap_char(string& s, const size_t& i, const size_t& j) {
        size_t num_swap = (j + 1 - i) / 2;
        for (size_t k = 0; k < num_swap; ++k) {
            swap(s[i + k], s[j - k]);
        }
    }
};

File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.rs
impl Solution {
    pub fn reverse_words(s: String) -> String {
        let mut i: usize = 0;
        let mut j: usize = 0;
        let size: usize = s.len();
        let mut s: Vec<u8> = s.into_bytes();

        fn swap_char(s: &mut Vec<u8>, start: &usize, end: &usize) {
            let num_swap: usize = (end + 1 - start) / 2;
            for k in 0..num_swap {
                s.swap(start + k, end - k);
            }
        }

        for k in 1..size {
            if (s[k] == (' ' as u8)) {
                swap_char(&mut s, &i, &j);
                i = k + 1;      // no need to update j because next char is guaranteed to update j
            } else {
                j = k;
            }
        }
        swap_char(&mut s, &i, &j);

        unsafe { String::from_utf8_unchecked(s) }
        // String::from_utf8(s).unwrap()
    }
}